GitUI/CommandsDialogs/FormCreateTag.cs

```diff
@@ -77,7 +77,7 @@ private string CreateTag()
                 File.WriteAllText(Path.Combine(Module.GetGitDirectory(), "TAGMESSAGE"), tagMessage.Text);
```

Comment:
I would move all tag-creation related functionality into the GitTagController - as the consumer of a method you don't need to know or care how your tag is created, all you have to do is to pass all necessary parameters and get your tag back.
Validation of the tag message and writing of the "TAGMESSAGE" file is the responsibility of the controller, not the caller.
The method should really look like this:

```
       private string CreateTag()
        {
            DialogResult = DialogResult.Cancel;

            string revision = commitPickerSmallControl1.SelectedCommitHash;
            if (revision.IsNullOrEmpty())
            {
                MessageBox.Show(this, _noRevisionSelected.Text, _messageCaption.Text);
                return string.Empty;
            }

            var s = _gitTagController.Tag(textBoxTagName.Text, revision, tagMessage.Text, ForceTag.Checked, annotate.SelectedIndex, textBoxGpgKey.Text);

            if (!string.IsNullOrEmpty(s))
                MessageBox.Show(this, s, _messageCaption.Text);

            if (s.Contains("fatal:"))
                return string.Empty;

            DialogResult = DialogResult.OK;
            return textBoxTagName.Text;
        }

```

I know it is the existing code and you haven't written it but I think it is a worthy refactor which is in line with your overall change.